app/vmagent/remotewrite/remotewrite.go

```diff
@@ -599,41 +677,25 @@ func tryShardingBlockAmongRemoteStorages(rwctxs []*remoteWriteCtx, tssBlock []pr
 			tmpLabels.Labels = hashLabels
 		}
 		h := getLabelsHash(hashLabels)
-		idx := h % uint64(len(shards))
-		i := 0
+
+		// Get the rwctxIdx through consistent hashing and then map it to the index in shards.
+		// The rwctxIdx is not always equal to the shardIdx, for example, when some rwctx are not available.
+		rwctxIdx := rwctxConsistentHashGlobal.GetNodeIdx(h, unhealthyIdx)
+		shardIdx := shardsIdxMap[rwctxIdx]
+
+		replicated := 0
 		for {
-			shards[idx] = append(shards[idx], ts)
-			i++
-			if i >= replicas {
+			shards[shardIdx] = append(shards[shardIdx], ts)
+			replicated++
+			if replicated >= replicas {
 				break
 			}
-			idx++
-			if idx >= uint64(len(shards)) {
-				idx = 0
+			shardIdx++
```

Comment:
This increment doesn't look safe. It's only possible to use next index from `healthyIdx` as target for replication.

Reply:
I don't get it, because len(shard) = len(healthyIdx). So next shard idx is actually point to the next healthy rwctx.